File: src/game/pieceGenerator.ts
import { Poly, Block } from "./poly";

export class PieceGenerator {

    public createPolyominoes(polySize: number): Poly[] {

        // create origin point
        let polys = new Array<Poly>();
        const block = new Block(0, 0);
        const startPoly = new Poly();
        startPoly.blocks.push(block);

        polys.push(startPoly);

        const hashes = new Set<String>();
        hashes.add(startPoly.getHash());

        for (let i = 1; i < polySize; i++) {
            polys = this.expandPolys(polys, hashes);
        }

        const hashPolys = new Set();
        const resultPolys = new Array<Poly>();
        for (let i = 0; i < polys.length; i++) {
            const poly = polys[i].clonePoly();
            if (poly.length == polySize) {
                this.normalisePoly(poly);
                const hash = poly.getHash();
                if (!hashPolys.has(hash)) {
                    hashPolys.add(hash);
                    resultPolys.push(poly);
                }
            }
        }
        resultPolys.forEach(poly => {
            this.centerPoly(poly);
        });
        return resultPolys;
    }

    private centerPoly(poly: Poly) {
        // find middle block

        let maxX = 0;
        let minX = 100000;
        let maxY = 0;
        let minY = 100000;

        poly.blocks.forEach(block => {
            if (block.x > maxX) {
                maxX = block.x;
            }
            if (block.x < minX) {
                minX = block.x;
            }
            if (block.y > maxY) {
                maxY = block.y;
            }
            if (block.y < minY) {
                minY = block.y;
            }
        });

        const middleX = Math.floor((maxX - minX) / 2);
        const middleY = Math.floor((maxY - minY) / 2);

        poly.blocks.forEach(block => {
            block.x -= middleX;
            block.y -= middleY;
        });
    }

    private normalisePoly(poly: Poly) {
        // find most negative x and y
        let negX = 0;
        let negY = 0;
        for (let i = 0; i < poly.length; i++) {
            if (poly.blocks[i].x < negX) {
                negX = poly.blocks[i].x;
            }
            if (poly.blocks[i].y < negY) {
                negY = poly.blocks[i].y;
            }
        }

        // add mod back to blocks
        for (let i = 0; i < poly.length; i++) {

            poly.blocks[i].x += Math.abs(negX);
            poly.blocks[i].y += Math.abs(negY);
        }

        // left align piece
        let smallX = poly.length;
        let smallY = poly.length;

        for (let i = 0; i < poly.length; i++) {
            if (poly.blocks[i].x < smallX) {
                smallX = poly.blocks[i].x;
            }
            if (poly.blocks[i].y < smallY) {
                smallY = poly.blocks[i].y;
            }
        }

        // add mod back to blocks
        for (let i = 0; i < poly.length; i++) {

            poly.blocks[i].x -= Math.abs(smallX);
            poly.blocks[i].y -= Math.abs(smallY);
        }
    }

    private attemptToGrowPoly(poly: Poly, block: Block, hashes: Set<String>, resultPolys: Set<Poly>): boolean {

        // check if block already exists in poly
        for (let i = 0; i < poly.length; i++) {
            // existing block
            const eB = poly.blocks[i];
            if (eB.x == block.x && eB.y == block.y) {
                return false;
            }
        }

        const blocks = Array.from(poly.blocks);
        blocks.push(block);
        let newPoly = Poly.fromBlocks(blocks);
        this.normalisePoly(newPoly);

        let addPoly = true;
        const newHashes = new Array<String>();
        for (let i = 0; i < 4; i++) {

            const hash = newPoly.getHash();
            if (hashes.has(hash)) {
                addPoly = false;
            } else {
                newHashes.push(hash);
                newPoly = newPoly.rotateClockwise();
                this.normalisePoly(newPoly);
            }
        }

        newHashes.forEach((hash, i, _newHashes) => {
            hashes.add(hash);
        });

        if (addPoly) {
            resultPolys.add(newPoly);
        }

        return addPoly;
    }

    private expandPolys(startPolys: Poly[], hashes: Set<String>): Poly[] {

        const resultPolys = new Set<Poly>(startPolys);

        // iterate through all polys
        for (let p = 0; p < startPolys.length; p++) {
            const poly = startPolys[p];

            const polyHash = poly.getHash();

            // iterate through all blocks in poly
            for (let i = 0; i < poly.length; i++) {

                // add a block in all cardinalities
                // left
                this.attemptToGrowPoly(poly, new Block(poly.blocks[i].x + 1, poly.blocks[i].y), hashes, resultPolys);

                // up
                this.attemptToGrowPoly(poly, new Block(poly.blocks[i].x, poly.blocks[i].y + 1), hashes, resultPolys);

                // right
                this.attemptToGrowPoly(poly, new Block(poly.blocks[i].x - 1, poly.blocks[i].y), hashes, resultPolys);

                // down
                this.attemptToGrowPoly(poly, new Block(poly.blocks[i].x, poly.blocks[i].y - 1), hashes, resultPolys);
            }
        }

        return Array.from(resultPolys);
    }

}

File: piece-generator/src/piece_generator.rs
use std::collections::HashSet;

use crate::poly::{Block, Poly};

pub fn create_polyominoes(poly_size: usize) -> Vec<Poly> {
    let mut polys = Vec::new();
    let block = Block::default();
    let mut start_poly = Poly::default();
    start_poly.blocks.push(block);
    polys.push(start_poly);

    for _ in 2..=poly_size {
        polys = create_polyominoes_from_previous(polys);
    }
    polys
}

pub fn create_polyominoes_from_previous(polys: Vec<Poly>) -> Vec<Poly> {
    let mut hashes = HashSet::new();
    expand_polys(polys, &mut hashes)
}

fn attempt_to_grow_poly(
    poly: &Poly,
    block: Block,
    hashes: &mut HashSet<String>,
    mut result_polys: HashSet<Poly>,
) -> HashSet<Poly> {
    // check if block already exists in poly
    for i in 0..poly.length() {
        // existing block
        let existing_block = &poly.blocks[i];
        if existing_block.x == block.x && existing_block.y == block.y {
            return result_polys;
        }
    }

    let mut blocks = poly.blocks.clone();
    blocks.push(block);
    let mut new_poly = Poly::from_blocks(blocks).normalise();

    let mut new_hashes = Vec::new();
    let mut add_poly = true;
    for i in 0..4 {
        let hash = new_poly.get_hash();
        if !hashes.contains(&hash) {
            new_hashes.push(hash);
        } else {
            add_poly = false;
            break;
        }

        if i != 3 {
            new_poly = new_poly.rotate_clockwise().normalise();
        }
    }

    if add_poly {
        result_polys.insert(new_poly);
        for h in new_hashes {
            hashes.insert(h);
        }
    }

    result_polys
}

fn expand_polys(start_polys: Vec<Poly>, hashes: &mut HashSet<String>) -> Vec<Poly> {
    let mut result_polys = HashSet::new();

    // iterate through all polys
    for poly in start_polys {
        // iterate through all blocks in poly
        for i in 0..poly.length() {
            // add a block in all cardinalities
            // left
            result_polys = attempt_to_grow_poly(
                &poly,
                Block::new(poly.blocks[i].x + 1, poly.blocks[i].y),
                hashes,
                result_polys,
            );

            // up
            result_polys = attempt_to_grow_poly(
                &poly,
                Block::new(poly.blocks[i].x, poly.blocks[i].y + 1),
                hashes,
                result_polys,
            );

            // right
            result_polys = attempt_to_grow_poly(
                &poly,
                Block::new(poly.blocks[i].x - 1, poly.blocks[i].y),
                hashes,
                result_polys,
            );

            // down
            result_polys = attempt_to_grow_poly(
                &poly,
                Block::new(poly.blocks[i].x, poly.blocks[i].y - 1),
                hashes,
                result_polys,
            );
        }
    }

    Vec::from_iter(result_polys)
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn poly_4_test() {
        assert_eq!(create_polyominoes(4).len(), 7);
    }

    #[test]
    fn poly_5_test() {
        assert_eq!(create_polyominoes(5).len(), 18);
    }

    #[test]
    fn poly_6_test() {
        assert_eq!(create_polyominoes(6).len(), 60);
    }

    #[test]
    fn poly_7_test() {
        assert_eq!(create_polyominoes(7).len(), 196);
    }

    #[test]
    fn poly_8_test() {
        assert_eq!(create_polyominoes(8).len(), 704);
    }

    #[test]
    fn poly_9_test() {
        assert_eq!(create_polyominoes(9).len(), 2500);
    }

    #[test]
    fn poly_12_test() {
        assert_eq!(create_polyominoes(12).len(), 126_759);
    }
}
